File: src/main/java/kitchenpos/product/tobe/application/ProductService.java
package kitchenpos.product.tobe.application;

import kitchenpos.common.tobe.Profanities;
import kitchenpos.product.tobe.domain.Product;
import kitchenpos.product.tobe.domain.ProductRepository;

import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;


import java.util.List;
import java.util.NoSuchElementException;

import java.util.UUID;

@Service
public class ProductService {
    private final ProductRepository productRepository;
    private final Profanities profanities;
    private final ApplicationEventPublisher applicationEventPublisher;


    public ProductService(
            final ProductRepository productRepository,
            final Profanities profanities,
            final ApplicationEventPublisher applicationEventPublisher) {
        this.productRepository = productRepository;
        this.profanities = profanities;
        this.applicationEventPublisher = applicationEventPublisher;
    }

    @Transactional
    public Product create(final Product request) {
        return productRepository.save(request);
    }

    @Transactional
    public Product changePrice(final UUID productId, final Product request) {
        final Product product = productRepository.findById(productId)
                .orElseThrow(NoSuchElementException::new);

        applicationEventPublisher.publishEvent(ProductPriceChangedEvent.of(productId, request.getProductPrice().getPrice()));

        product.updatePrice(request.getProductPrice());
        return product;
    }

    @Transactional(readOnly = true)
    public List<Product> findAll() {
        return productRepository.findAll();
    }
}

File: src/main/kotlin/kitchenpos/product/tobe/application/ProductService.kt
package kitchenpos.product.tobe.application

import java.util.*
import kitchenpos.common.domain.Profanities
import kitchenpos.product.tobe.application.dto.ChangeProductPriceReq
import kitchenpos.product.tobe.application.dto.CreateProductReq
import kitchenpos.product.tobe.application.dto.ProductResp
import kitchenpos.product.tobe.domain.Product
import kitchenpos.product.tobe.domain.ProductName
import kitchenpos.product.tobe.domain.ProductNamePolicy
import kitchenpos.product.tobe.domain.ProductPrice
import kitchenpos.product.tobe.domain.ProductRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service("tobeProductService")
class ProductService(
    private val productRepository: ProductRepository,
    private val profanities: Profanities,
) {

    @Transactional
    fun create(request: CreateProductReq): ProductResp {
        val product = productRepository.save(
            Product(
                productName = ProductName(ProductNamePolicy(profanities), request.name),
                productPrice = ProductPrice(request.price)
            )
        )
        return ProductResp.of(product)
    }

    fun findAll(): List<ProductResp> {
        return productRepository.findAll().map(ProductResp::of)
    }

    @Transactional
    fun changePrice(productId: UUID, request: ChangeProductPriceReq) {
        val product =
            productRepository.findById(productId).orElseThrow { throw NoSuchElementException("상품을 찾을 수 없습니다.") }
        product.changePrice(ProductPrice(request.price))
        productRepository.save(product)
    }
}
